### ISSUE TITLE
lxd storage buckets: Buckets are missing project name prefix

### ISSUE THREAD

**Issue by MusicDin** [lxc/lxd#11828]
## TLDR 

When buckets are created they are missing a project name prefix. (`buckets/mybucket` -> `buckets/default_mybucket`)

---

## Example

For example, we have a pool named `mypool` and two projects: `default` and `myproject`. 

We can create two volumes with the same name in two different projects.

```sh
lxc storage volume create mypool myvol --project default
lxc storage volume create mypool myvol --project myproject

lxc storage volume list --all-projects --format compact
#  PROJECT   TYPE   NAME   DESCRIPTION  CONTENT-TYPE  USED BY  
#  default  custom  myvol               filesystem    0        
#  new      custom  myvol               filesystem    0      
```

This does not apply to the buckets:
```sh
lxc storage bucket create pool mybucket --project default
# Storage bucket mybucket created

lxc storage bucket create pool mybucket --project myproject
# Error: Failed creating storage bucket: Failed inserting storage 
# bucket "mybucket" for project "myproject" in pool "pool" into 
# database: A bucket for that name already exists
```

## Recovery

Besides not being able to create buckets with the same name in different projects, the missing prefix makes extracting the project name during the recovery process of buckets impossible (also see issue #11704).

## Suggestions

@stgraber suggested that we could patch bucket names on LXD upgrade, and that we could also introduce `backup.yaml` into buckets to be able to restore access keys.

**Comment by tomponline** [comment]
This is by design because the bucket name is globally unique.

**Comment by tomponline** [comment]
We have a patch system for making storage level changes we can go through for performing the rename.

**Comment by MusicDin** [comment]
Should we still prefix them to be able to recover the project name?
Maybe we could store the project name in the `backup.yaml` file, and be able to recover both the project name and the access keys?

**Comment by tomponline** [comment]
I dont think we need a backup.yaml as part of the scope of restoring the bucket at least, as these can be re-added manually, the important part is getting access to the S3 bucket back. I'm not ruling it out, but if we do it, we should do it for custom volumes too, so its quite a bit more involved.

**Comment by tomponline** [comment]
> Should we still prefix them to be able to recover the project name?

Yep I think we could do that indeed.

**Comment by MusicDin** [comment]
Yes, but it may be convenient to also restore access keys as they are most likely used across different applications, so it would reduce the effort from the user?

I think that `backup.yaml` can be stored in bucket, since minio uses a subdirectory `minio` for its data.

**Comment by stgraber** [comment]
Re-opening the issue as we should indeed be adding the prefix to put the buckets in line with volumes as far as storage is concerned. The fact that they must have a unique name across the LXD deployment/cluster is unrelated, having the project name encoded in the underlying storage is a common pattern by now in LXD and we should follow it there as it will make restoration much easier.

**Comment by stgraber** [comment]
Ah, I see, we just want to make a long todo list in the feature issue, we can do that :)

**Comment by tomponline** [comment]
Yes please cheers
